### PR TITLE
[charlie] Todoresources

### PR DESCRIPTION
**Pull request by MelindaK**

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -29,6 +29,8 @@ def handleBaseException(error):
 from app.todo_lists.models import TodoList
 from app.todos.models import Todo
 from app.users.resources import UserListResource, UserResource
+from app.todos.resources import todoMultiResource, todoResource
```
Comment (cfierro) [inline app/__init__.py]: Class names should use `CapWords` convention instead of `camelCase`. s/todoMultiResource/TodoMultiResource and s/todoResource/TodoResource

File: app/__init__.py [hunk 2]
```diff
@@ -29,6 +29,8 @@ def handleBaseException(error):
...
+from app.todo_list.resources import todoListMultiResource, todoListResource
```
Comment (cfierro) [inline app/__init__.py]: Same comment as above.

File: app/__init__.py [hunk 3]
```diff
@@ -40,3 +42,11 @@ def handleBaseException(error):
 # All resources that need to be routed should be routed here.
 api.add_resource(UserListResource, '/users', '/users/')
 api.add_resource(UserResource, '/users/<int:userId>', '/users/<int:userId>/')
+api.add_resource(todoMultiResource, '/todos', '/todos/')
+api.add_resource(todoResource, '/todos/<int:todoId>', 'todos/<int:todoId>/')
+api.add_resource(todoListMultiResource, '/todolist', '/todolist/')
+api.add_resource(todoListResource, '/todolist/<int:todoListId>',
+                 '/todolist/<int:todoListId>/')
+
+db.create_all()
```
Comment (cfierro) [inline app/__init__.py]: There is already a `db.create_all()` on line 35. You can delete this one here.

File: app/lib/models.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+from app import db
+
+
+class Base(db.Model):
+
```
Comment (cfierro) [inline app/lib/models.py]: Let's add a small comment here explaining all classes should inherit from this model or that it provides common fields for all models.

File: app/lib/models.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+from app import db
+
+
+class Base(db.Model):
```
Comment (cfierro) [inline app/lib/models.py]: Also, instead of calling this `Base` lets call this `SomethingMixin`. Totally my bad earlier, but after looking at the documentation for sqlalchemy a bit more closely, this seems like the better way to do things: http://docs.sqlalchemy.org/en/rel_1_0/orm/extensions/declarative/mixins.html

Hence we need to change several things:
1) Our app's `__init__.py` file should perhaps make a variable called Base that makes the declarative base. https://pythonhosted.org/Flask-SQLAlchemy/api.html#flask.ext.sqlalchemy.SQLAlchemy.make_declarative_base
2) This class should be a "Mixin"
3) Whenever we want to use our Mixin, we should declare a class with both `Base` and `SomethingMixin`. It looks like the ordering doesn't matter, but lets be consistent.

File: app/__init__.py [hunk 6]
```diff
@@ -40,3 +42,11 @@ def handleBaseException(error):
 # All resources that need to be routed should be routed here.
 api.add_resource(UserListResource, '/users', '/users/')
 api.add_resource(UserResource, '/users/<int:userId>', '/users/<int:userId>/')
+api.add_resource(todoMultiResource, '/todos', '/todos/')
+api.add_resource(todoResource, '/todos/<int:todoId>', 'todos/<int:todoId>/')
```
Comment (cfierro) [inline app/__init__.py]: I was getting an error because the second url did not start with a `/`

File: app/__init__.py [hunk 7]
```diff
@@ -40,3 +42,11 @@ def handleBaseException(error):
...
+api.add_resource(todoListMultiResource, '/todolist', '/todolist/')
+api.add_resource(todoListResource, '/todolist/<int:todoListId>',
+                 '/todolist/<int:todoListId>/')
```
Comment (cfierro) [inline app/__init__.py]: Could you make the urls `todolists` instead.

File: app/permissions/models.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
+from app import db
+
+from app.lib.models import Base
+
+
+class TodoListPermission(Base):
+    userId = db.Column(db.Integer, db.ForeignKey('user.id'))
+    todoListId = db.Column(db.Integer, db.ForeignKey('todoList.id'))
```
Comment (cfierro) [inline app/permissions/models.py]: I think the foreign key should be `todo_list.id`

File: app/permissions/models.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
+from app import db
+
+from app.lib.models import Base
+
+
+class TodoListPermission(Base):
```
Comment (cfierro) [inline app/permissions/models.py]: Use the Base and Mixin here

File: app/todo_lists/models.py [hunk 10]
```diff
@@ -1,14 +1,14 @@
...
-    id = db.Column(db.Integer, primary_key=True)
+
+class TodoList(Base):
     name = db.Column(db.String)
-    userId = db.Column(db.Integer, db.ForeignKey('user.id'), primary_key=True)
-    user = db.relationship('User',
-        backref=db.backref('lists', lazy='dynamic'))
+    creatorId = db.Column(db.Integer, db.ForeignKey('user.id'))
+    creator = db.relationship(
+        'User', backref=db.backref('lists', lazy='dynamic'))
```
Comment (cfierro) [inline app/todo_lists/models.py]: Just a reminder that we will just use the default `select` here.

File: app/todo_lists/resources.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+from app import db, TodoList
+from app.lib.response_util import buildOkResponse
+
+
+def _returnTodoList(todoList):
+    """Private method to convert a todoList object into a dictionary.
+    """
+    return {
+        "id": todoList.id,
+        "name": todoList.name
```
Comment (cfierro) [inline app/todo_lists/resources.py]: Do you think we should also return the list creator/owner name?

File: app/__init__.py (hunk 7 above)
Comment (cfierro) [inline app/__init__.py]: I typically see the resources of a restful API be plural, but really consistency is all that matters.

File: app/permissions/models.py (hunk 9 above)
Comment (cfierro) [inline app/permissions/models.py]: Also please add a doc string.

File: app/permissions/models.py [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+    userId = db.Column(db.Integer, db.ForeignKey('user.id'))
+    todoListId = db.Column(db.Integer, db.ForeignKey('todoList.id'))
+
+    def __init__(self, userId, todoListId):
+        self.userId = userId
```
Comment (cfierro) [inline app/permissions/models.py:22]: Please add a doc string here too. It can be super simple too like

```
"""Constructor.

Args:
    userId - ID of user allowed to view the given todo list.
    todoListId - ID of todo list being granted access to.
"""
```

File: app/todo_lists/models.py [hunk 13]
```diff
@@ -1,14 +1,14 @@
 from app import db
 
+from app.lib.models import Base
 
-class TodoList(db.Model):
-    id = db.Column(db.Integer, primary_key=True)
+
+class TodoList(Base):
```
Comment (cfierro) [inline app/todo_lists/models.py]: Same comment about using the declared base and your mixin here. Also please add a small doc string.

File: app/todo_lists/models.py [hunk 14]
```diff
@@ -1,14 +1,14 @@
...
-    userId = db.Column(db.Integer, db.ForeignKey('user.id'), primary_key=True)
-    user = db.relationship('User',
-        backref=db.backref('lists', lazy='dynamic'))
+    creatorId = db.Column(db.Integer, db.ForeignKey('user.id'))
+    creator = db.relationship(
+        'User', backref=db.backref('lists', lazy='dynamic'))
 
-    def __init__(self, name, userId, id=None):
-        self.id = id
+    def __init__(self, name, creatorId):
```
Comment (cfierro) [inline app/todo_lists/models.py:15]: Same comment about adding a small doc string.

File: app/permissions/models.py (hunk 9 above)
Comment (cfierro) [inline app/permissions/models.py]: `userId` and `todoListId` should be enough to grant sharing permissions, but if you would like to be fancy, we could add like a "role" column that lets you share to someone but grant explicit permissions like "view", "modify", and "admin".

File: app/todo_lists/resources.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+class TodoListMultiResource(Resource):
+    """Class for creating and accessing lists.
+    """
+    def get(self):
+        """Method gets a list of todoLists and returns them in an Ok response.
+        """
+        todoLists = TodoList.query.all()
+
+        return buildOkResponse([_returnTodoList(todoList) for todoList in
+                               todoLists])
```
Comment (cfierro) [inline app/todo_lists/resources.py]: Small nit. `todoLists` should start right underneath the above `_` (unless your linter is telling you otherwise).

```
return buildOkResponse([_returnTodoList(todoList) for todoList in
                        todoLists])
```

File: app/todo_lists/resources.py [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+    def post(self):
+        """Method creates a todoList and returns it in an Ok response.
+        """
+        newListData = json.loads(request.form['data'])
```
Comment (cfierro) [inline app/todo_lists/resources.py]: Just a reminder that we are just going to pull values from the form directly instead of embedding it in the `data` key.

Comment (cfierro) [inline app/todo_lists/resources.py]: Please also add a little validation to the form data given and raise any exceptions you think are necessary.

File: app/todo_lists/resources.py [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+
+class TodoListResource(Resource):
+    """Class to get, update, or delete a single todoList.
+    """
+    def put(self, listId):
+        """Method that updates and returns a todoList in an Ok response.
+        """
+        todoList = TodoList.query.get(listId)
+        todoListData = json.loads(request.form['data'])
```
Comment (cfierro) [inline app/todo_lists/resources.py]: Same comment above about just using `request.form` directly.

File: app/todo_lists/resources.py [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+        db.session.commit()
+
+        return buildOkResponse(_returnTodoList(newList))
+
+
+class TodoListResource(Resource):
+    """Class to get, update, or delete a single todoList.
+    """
+    def put(self, listId):
+        """Method that updates and returns a todoList in an Ok response.
```
Comment (cfierro) [inline app/todo_lists/resources.py:52]: listId is pretty self explanatory, but lets be consistent and add an Args section

File: app/todo_lists/resources.py [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+        todoList.name = todoListData.get('name') or todoList.name
+        todoList.userId = todoListData.get('userId') or todoList.userId
+
+        db.session.update()
+
+        return buildOkResponse(_returnTodoList(todoList))
+
+    def get(self, listId):
+        """Method that gets and returns a todoList in an Ok response.
+        """
```
Comment (cfierro) [inline app/todo_lists/resources.py:81]: Same comment as above about including an args section.

File: app/todo_lists/resources.py [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+        todoList = TodoList.query.get(listId)
```
Comment (cfierro) [inline app/todo_lists/resources.py]: There is a possible exception in this method. We can talk about it on Monday, but if you can track it down, validate for it, and raise an exception as necessary, that would be great!

File: app/todo_lists/resources.py [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+        return buildOkResponse(_returnTodoList(todoList))
+
+    def delete(self, listId):
+        """Method that deletes a todoList and returns result None.
+        """
+        todoList = TodoList.query.get(listId)
+        db.session.delete(todoList)
```
Comment (cfierro) [inline app/todo_lists/resources.py:117]: Similar comments as in the put and get methods.

File: app/todos/models.py [hunk 22]
```diff
@@ -1,8 +1,9 @@
 from app import db
 
+from app.lib.models import Base
 
-class Todo(db.Model):
-    id = db.Column(db.Integer, primary_key=True)
+
+class Todo(Base):
```
Comment (cfierro) [inline app/todos/models.py]: Same comments as above about using the declarative base, your mixin, and adding doc strings.

File: app/todos/resources.py [hunk 23]
```diff
@@ -0,0 +1,85 @@
+# Ask about getting branches up to date once committed
```
Comment (cfierro) [inline app/todos/resources.py]: Hopefully I answered this on Friday :), but if you need a reminder, the general process is:
1) Bring you local master branch up to date `git pull origin master`
2) Merge your master branch to your working branch `git merge master`
3) Resolve any conflicts that pop up.
4) Add and commit

File: app/todos/resources.py [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+
+def _returnTodo(todo):
+    """Private method to convert a todo object into a dictionary.
+
+    Args:
+        todo - A todo model.
+    """
+    return {
+            "id": todo.id,
+            "subject": todo.subject
```
Comment (cfierro) [inline app/todos/resources.py]: Do you think we should return more information than this? Imagine that you are the browser and want to display to the user a todo item. What information will you need?

File: app/todos/resources.py [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+
+class todoMultiResource(Resource):
```
Comment (cfierro) [inline app/todos/resources.py]: Class names should start with a capital letter.

File: app/todos/resources.py [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+    """Class for creating and accessing todos.
+    """
+    def get(self):  # Should this take a user and return their todos?
```
Comment (cfierro) [inline app/todos/resources.py]: Would you like to implement sessions, cookies, and hashed passwords now, or sort of hack around this by adding a userId query parameter to each request? I'm glad you noticed that this is a problem. But we can fix it!

File: app/todos/resources.py [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+        return buildOkResponse([_returnTodo(todo) for todo in todos])
+
+    def post(self):
+        """Method adds a new todo.
+        """
+        newTodoData = json.loads(request.form['data'])  # this is a dictionary
+        newTodo = Todo(newTodoData['subject'], newTodoData['todoListId'],
+                       newTodoData['dueDate'], newTodoData['description'],
+                       newTodoData['creatorId'], newTodoData['priority'],
+                       newTodoData['completed'], newTodoData['assigneeId'])
```
Comment (cfierro) [inline app/todos/resources.py]: Same comments as above about using `request.form` directly and adding any exception handling.

File: app/todos/resources.py [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+        db.session.add(newTodo)
+        db.session.commit()
+
+        return buildOkResponse(_returnTodo(newTodo))
+
+
+class todoResource(Resource):
```
Comment (cfierro) [inline app/todos/resources.py]: Same comment as above about starting classes with a capital letter.

File: app/todos/resources.py [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+            todoId - Interger, primary key identifying the todo.
+        """
+        todo = Todo.query.get(todoId)
+        todoData = json.loads(request.form['data'])
+
+        todo.subject = todoData.get('subject') or todo.subject
+        todo.dueDate = todoData.get('dueDate') or todo.dueDate
+        todo.description = todoData.get('description') or todo.description
+        todo.priority = todoData.get('priority') or todo.priority
+        todo.completed = todoData.get('completed') or todo.completed
```
Comment (cfierro) [inline app/todos/resources.py]: Similar comment as above about using `request.form` directly and adding any exception handling.

File: app/todos/resources.py [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+
+        db.session.commit()
+
+        return buildOkResponse(_returnTodo(todo))
+
+    def get(self, todoId):
+        """Method gets and returns a single todo in an OK response.
+        """
+        todo = Todo.query.get(todoId)
+        return buildOkResponse(_returnTodo(todo))
```
Comment (cfierro) [inline app/todos/resources.py]: Similar comment as above about there being a possible exception here.

File: app/todos/resources.py [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+
+    def delete(self, todoId):
+        """Method deletes a todo and returns result none.
+        """
+        todo = Todo.query.get(todoId)
+        db.session.delete(todo)
+        db.session.commit()
```
Comment (cfierro) [inline app/todos/resources.py:151]: Same comment as above.

File: app/users/models.py [hunk 32]
```diff
@@ -1,7 +1,8 @@
 from app import db
+from app.lib.models import Base
 
-class User(db.Model):
-    id = db.Column(db.Integer, primary_key=True)
+
+class User(Base):
```
Comment (cfierro) [inline app/users/models.py]: Similar comments about using the declared base, your mixin, and adding doc strings.

**Comment by cfierro** [comment]
@MelindaK done with comments. Quite a few, but they should go by pretty quickly. Great work! There are only a few steps left, and then we can get started on the front end side of this project.

File: app/lib/models.py (hunk 5 above)
Comment (cfierro) [inline app/lib/models.py]: Just adding another comment to explain the bug. Our mixin should still inherit from db.Model, and we now need to define the table name explicitly. We can do so in two ways
1) In our mixin have a function that auto does it for us: `def __tablename__(cls)`
2) On every table defined `__tablename__ = <table name>`

**Comment by MelindaK** [comment]
Addressed PR comments and added unit tests for Todo and TodoList

**Comment by MelindaK** [comment]
Finished todo/todolist resources adding permissions and auth

File: app/__init__.py [hunk 33]
```diff
@@ -24,12 +24,15 @@ def handleBaseException(error):
...
 db = SQLAlchemy(app)
+Base = db.make_declarative_base()
 
 # Import all models here for easier importing.
 from app.users.models import User
 from app.todo_lists.models import TodoList
 from app.todos.models import Todo
+from app.permissions.models import TodoListPermission
 
+# db.drop_all()
```
Comment (cfierro) [inline app/__init__.py]: You can delete this here.

File: app/lib/status.py [hunk 34]
```diff
@@ -34,6 +34,28 @@ def __init__(self, message='Bad request', details=None, result=None,
                                          result=result, info=info)
 
 
+class Unauthorized(TodoException):
+    """Exception where unauthorized request is made.
```
Comment (cfierro) [inline app/lib/status.py]: Could you rephrase this to be `Exception raised when an unauthorized request is made.`

File: app/lib/status.py [hunk 35]
```diff
@@ -34,6 +34,28 @@ def __init__(self, message='Bad request', details=None, result=None,
...
+    statusCode = 401
+
+    def __init__(self, message="Unauthorized request", details=None,
+                 result=None, info=None):
+        super(Unauthorized, self).__init__(message, details=None, result=None,
+                                           info=None)
+
+
+class Forbidden(TodoException):
+    """Exception where permission is denied.
```
Comment (cfierro) [inline app/lib/status.py]: Similar comment as above.

File: app/permissions/models.py [hunk 36]
```diff
@@ -0,0 +1,23 @@
...
+from app.lib.models import TableMixin
+
+
+class TodoListPermission(TableMixin, Base):
+    """Constructor.
+
+    Args:
+        userId - ID of user allowed to view the given todo list.
+        todoListId - ID of todo list being granted access to.
+    """
```
Comment (cfierro) [inline app/permissions/models.py]: This doc string should actually go into the `__init__` method

File: app/todo_lists/models.py [hunk 37]
```diff
@@ -1,14 +1,29 @@
...
+        'User', backref=db.backref('lists'))
 
-    def __init__(self, name, userId, id=None):
-        self.id = id
+    def __init__(self, name, creatorId):
         self.name = name
-        self.userId = userId
+        self.creatorId = creatorId
+
+        # TodoListPermission(self.creatorId, self.id)
```
Comment (cfierro) [inline app/todo_lists/models.py]: Delete this comment.

File: app/todo_lists/resources.py [hunk 38]
```diff
@@ -0,0 +1,123 @@
...
+        """Method that updates and returns a todoList in an Ok response.
+
+        Args:
+            listId - An integer, primary key that identifies the todo list.
+        """
+        userId = session.get('userId')
+        todoList = TodoList.query.get(todoListId)
+        permission = TodoList.query.filter()
+
+        TodoListPermission.query.filter_by(userId=userId)
```
Comment (cfierro) [inline app/todo_lists/resources.py]: I think you can get rid of this and line 59. You redefine permission below on 66.

File: app/todos/models.py [hunk 39]
```diff
@@ -1,28 +1,49 @@
...
+                'subject': self.subject,
+                'todoListId': self.todoListId,
+                'creatorId': self.creatorId,
+                'dueDate': self.dueDate,
+                'description': self.description,
+                'priority': self.priority,
+                'completed': self.completed,
+                'assignee': assigneeName,
+                'list': self.todoList.name
+            }
```
Comment (cfierro) [inline app/todos/models.py]: Nit. You can unindent 39-49 by one tab.

File: app/todo_lists/resources.py [hunk 40]
```diff
@@ -0,0 +1,123 @@
...
+
+        if todoList is None:
+            raise status.NotFound()
+
+        permission = TodoListPermission.query.filter(
+                           and_(TodoListPermission.todoListId == todoList.id,
```
Comment (cfierro) [inline app/todo_lists/resources.py]: Let's not indent this and_ so much. Usually we just indent once when continuing in a new line and not in an indented context. But if your linter is telling you otherwise, then no problem!

File: app/todo_lists/resources.py [hunk 41]
```diff
@@ -0,0 +1,123 @@
...
+            todoListId - An integer, primary key that identifies the todo list.
+        """
+        userId = session.get('userId')
+        todoList = TodoList.query.get(todoListId)
+
+        if todoList is None:
+            raise status.NotFound()
+
+        permission = TodoListPermission.query.filter(
+                           and_(TodoListPermission.todoListId == todoList.id,
```
Comment (cfierro) [inline app/todo_lists/resources.py]: Same comment about not indenting this so much.

File: app/todo_lists/resources.py [hunk 42]
```diff
@@ -0,0 +1,123 @@
...
```
Comment (cfierro) [inline app/todo_lists/resources.py]: Same comment about not indenting this so much.

File: app/todos/resources.py [hunk 43]
```diff
@@ -0,0 +1,153 @@
...
+class TodoMultiResource(Resource):
+    """Class for creating and accessing todos.
+    """
+    @authentication.requiresAuth
+    def get(self):
+        """Method gets a list of all the todos and returns them in an OK
+        response.
+        """
+        userId = session.get('userId')
+        permissions = TodoListPermission.query.filter_by(userId=userId)
```
Comment (cfierro) [inline app/todos/resources.py]: You should move this query into the if block below. Only when todoListId is None should we actually execute this query. DB queries are slow compared to other typical actions.

File: app/todos/resources.py [hunk 44]
```diff
@@ -0,0 +1,153 @@
...
+from app.lib import authentication, response_util, status
+
+
+class TodoMultiResource(Resource):
+    """Class for creating and accessing todos.
+    """
+    @authentication.requiresAuth
+    def get(self):
+        """Method gets a list of all the todos and returns them in an OK
+        response.
```
Comment (cfierro) [inline app/todos/resources.py]: Let's add to the comment and explain that the API call can be filtered by todoListId.

File: app/todos/resources.py [hunk 45]
```diff
@@ -0,0 +1,153 @@
...
+            if not permissions:
+                raise status.Forbidden()
+
+            todos = Todo.query.filter_by(todoListId=todoListId)
+
+            return response_util.buildOkResponse([todo.toDict() for todo in todos])
+
+    @authentication.requiresAuth
+    def post(self):
+        """Method adds a new todo.
```
Comment (cfierro) [inline app/todos/resources.py]: For consistency with your other comments lets add "... and returns the todo in an OK response."

File: app/todos/resources.py [hunk 46]
```diff
@@ -0,0 +1,153 @@
...
+        """
+        userId = session.get('userId')
+        todoListId = request.form.get('todoListId')
+
+        if not(request.form.get('subject') and request.form.get('todoListId')):
```
Comment (cfierro) [inline app/todos/resources.py]: You can replace request.form.get('todoListId') with the variable todoListId that you defined above.

File: app/todos/resources.py [hunk 47]
```diff
@@ -0,0 +1,153 @@
...
+
+        permission = TodoListPermission.query.filter(
+                           and_(TodoListPermission.todoListId == todoListId,
+                                TodoListPermission.userId == userId)).first()
+
+        if permission is None:
+            raise status.Forbidden()
+
+        todo = Todo(request.form.get('subject'),
+                    request.form.get('todoListId'),
```
Comment (cfierro) [inline app/todos/resources.py]: Same comment as above.

File: app/todos/resources.py [hunk 48]
```diff
@@ -0,0 +1,153 @@
...
+                    request.form.get('completed'),
+                    request.form.get('assigneeId'))
+        db.session.add(todo)
+        db.session.commit()
+
+        return response_util.buildOkResponse(todo.toDict())
+
+
+class TodoResource(Resource):
+    """Class for updating, delelting, getting single todo.
```
Comment (cfierro) [inline app/todos/resources.py]: s/delelting/deleting

Comment (cfierro) [inline app/todos/resources.py]: getting a\* single todo

File: app/todos/resources.py [hunk 49]
```diff
@@ -0,0 +1,153 @@
...
+    """
+    @authentication.requiresAuth
+    def put(self, todoId):
+        """Method updates a todo information and returns todo in an OK response.
```
Comment (cfierro) [inline app/todos/resources.py]: s/information/instance

Comment (cfierro) [inline app/todos/resources.py]: and returns it\* in

**Comment by MelindaK** [comment]
Addressed all comments

**Comment by cfierro** [comment]
LGTM just needs a drop and create all on the database.

After dropping all the tables you can bring back all your data by doing: 

```
INSERT INTO "user" VALUES(1,NULL,NULL,'Melinda Kilner','mhkilner@gmail.com','password');
INSERT INTO "todo_list" VALUES(1,NULL,NULL,'Test List',1);
INSERT INTO "todo_list_permission" VALUES(1,NULL,NULL,1,1);
INSERT INTO "todo" VALUES(1,NULL,NULL,'Test Task Item',1,1,NULL,'Do the task.',1,0,1);
```

The way I figured that out is by doing `sqlite3 app.db .dump > app_dump`

**Comment by cfierro** [comment]
LGTM

**Comment by cfierro** [comment]
:+1:
